### PR TITLE
Optimize causal mask using torch.where

### PR DESCRIPTION
**Pull request by Akababa**
Instead of multiplying by 1.0 float mask, use torch.where with a bool mask for increased performance.

### CODE REVIEW

**Comment by julien-c** [comment]
Thanks – what's the PyTorch compatibility on this?

**Comment by Akababa** [comment]
Not sure about that, where can I find more info on compatibility? I think it only relies on torch.where (introduced <= 1.0.0) and tensors of dtype torch.bool (introduced in 1.2.0). Does the None (newaxis) slicing introduce compatibility issues?

**Comment by nsmetanin** [comment]
Hi, I'd recommend to make the following changes:
1. Keep the original shapes of _bias_ buffer (because otherwise it breaks loading of already trained models) and make dtype equal to torch.uint8, so it'd be compatible with pytorch 1.0.0 as no torch.bool type available.
`self.register_buffer("bias", torch.tril(torch.ones((n_ctx, n_ctx), dtype=torch.uint8)).view(1, 1, n_ctx, n_ctx))`
2. Keep -1e4 constant in a buffer to reduce allocations on each _attn call:
`self.register_buffer("masked_bias", torch.tensor(-1e4))`
3. Keep `b = self.bias[:, :, ns - nd : ns, :ns]` line as _bias_ buffer have the original shape now
4. So the _where_ statement should look like `w = torch.where(b, w, self.masked_bias)`

As a result, overall speedup will be at 10-15% here as I measured.

**Comment by patrickvonplaten** [comment]
Hi @Akababa, 

Thanks for the PR. I think this is a great change. I checked and it does lead to a significant speed-up :-) 

Could you fix the tests and I think then we can merge (see https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md)

1) You should fetch the master branch and rebase your branch on top of it.
2) Make sure to run `make style` in the root folder before pushing to pass the "check_code_quality" test.

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Great work @Akababa - this looks good to me! 

@LysandreJik @thomwolf - could you check and merge?

**Review by LysandreJik — APPROVED** [review]
LGTM

**Comment by patrickvonplaten** [comment]
Checked slow hardcoded GPT2 tests and it looks all good!
